pkg/controller/nodeipam/ipam/multicidrset/multi_cidr_set.go

```diff
@@ -0,0 +1,365 @@
+	// nodeMask is the network mask assigned to the nodes.
+	nodeMask net.IPMask
+	// allocatedCIDRs counts the number of CIDRs allocated.
+	allocatedCIDRs int
+	// nextCandidate points to the next CIDR that should be free.
+	nextCandidate int
+}
+
+// ClusterCIDR is an internal representation of the ClusterCIDR API object.
+type ClusterCIDR struct {
+	// Name of the associated ClusterCIDR API object.
+	Name string
+	// IPv4CIDRSet is the MultiCIDRSet representation of ClusterCIDR.spec.ipv4
+	// of the associated ClusterCIDR API object.
+	IPv4CIDRSet *MultiCIDRSet
+	// IPv6CIDRSet is the MultiCIDRSet representation of ClusterCIDR.spec.ipv6
+	// of the associated ClusterCIDR API object.
+	IPv6CIDRSet *MultiCIDRSet
+	// AssociatedNodes is used to identify which nodes have CIDRs allocated from this ClusterCIDR.
+	// Stores a mapping of node name to association status.
+	AssociatedNodes map[string]bool
+	// Terminating is used to identify whether ClusterCIDR has been marked for termination.
+	Terminating bool
+}
+
+const (
+	// The subnet mask size cannot be greater than 16 more than the cluster mask size
+	// TODO: https://github.com/kubernetes/kubernetes/issues/44918
+	// clusterSubnetMaxDiff limited to 16 due to the uncompressed bitmap.
+	// Due to this limitation the subnet mask for IPv6 cluster cidr needs to be >= 48
+	// as default mask size for IPv6 is 64.
+	clusterSubnetMaxDiff = 16
+	// halfIPv6Len is the half of the IPv6 length.
+	halfIPv6Len = net.IPv6len / 2
+)
+
+// CIDRRangeNoCIDRsRemainingErr is an error type used to denote there is no more
+// space to allocate CIDR ranges from the given CIDR.
+type CIDRRangeNoCIDRsRemainingErr struct {
+	// CIDR represents the CIDR which is exhausted.
+	CIDR string
+}
+
+func (err *CIDRRangeNoCIDRsRemainingErr) Error() string {
+	return fmt.Sprintf("CIDR allocation failed; there are no remaining CIDRs left to allocate in the range %s", err.CIDR)
+}
+
+// CIDRSetSubNetTooBigErr is an error type to denote that subnet mask size is too
+// big compared to the CIDR mask size.
+type CIDRSetSubNetTooBigErr struct {
+	cidr            string
+	subnetMaskSize  int
+	clusterMaskSize int
+}
+
+func (err *CIDRSetSubNetTooBigErr) Error() string {
+	return fmt.Sprintf("Creation of New CIDR Set failed for %s. "+
+		"PerNodeMaskSize %d is too big for CIDR Mask %d, Maximum difference allowed "+
+		"is %d", err.cidr, err.subnetMaskSize, err.clusterMaskSize, clusterSubnetMaxDiff)
+}
+
+// NewMultiCIDRSet creates a new MultiCIDRSet.
+func NewMultiCIDRSet(cidrConfig *net.IPNet, perNodeHostBits int) (*MultiCIDRSet, error) {
+	clusterMask := cidrConfig.Mask
+	clusterMaskSize, bits := clusterMask.Size()
+
+	var subNetMaskSize int
+	switch /*v4 or v6*/ {
+	case cidrConfig.IP.To4() != nil:
+		{
+			subNetMaskSize = 32 - perNodeHostBits
+		}
+	case cidrConfig.IP.To16() != nil:
+		{
+			subNetMaskSize = 128 - perNodeHostBits
+		}
+	}
+
+	if (cidrConfig.IP.To4() == nil) && (subNetMaskSize-clusterMaskSize > clusterSubnetMaxDiff) {
```

Comment:
https://github.com/kubernetes/kubernetes/pull/109090/commits/7c3f82e96bd42ca1038722b5366e1442f0c22a78#r934767400

To4() is always nil for IPv6 :shrug:

Reply:
Borrowed from https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodeipam/ipam/cidrset/cidr_set.go#L83

I think it's trying to check whether the address is IPv6, updated it to` if netutils.IsIPv6(cidrConfig.IP)`

Reply:
Seems like this wasn't done?

Reply:
Done, was added to a different commit